tools/themis/src/utils.rs

```diff
@@ -0,0 +1,123 @@
+        .filter(|package| metadata.workspace_members.contains(&package.id))
+        .map(|package| {
+            let raw = parse_toml(&package.manifest_path)?;
+            Ok(Package { parsed: package, raw })
+        })
+        .collect::<anyhow::Result<Vec<Package>>>()?;
+
+    Ok(Workspace { root: metadata.workspace_root, members })
+}
+
+/// returns true if the rule failed, false if otherwise
+pub fn check_and_report(outcome: Result<(), Error>, workspace: &Workspace) -> anyhow::Result<bool> {
+    match outcome {
+        Err(Error::RuntimeError(err)) => Err(err),
+        Err(Error::OutcomeError { msg, expected, outliers }) => {
+            let mut report = format!(
+                "{c_heading}(i) {}:{c_none}{}",
+                msg,
+                expected.map_or("".to_string(), |Expected { value, reason }| format!(
+                    " [expected: {c_expected}{}{c_none}{}]",
+                    value,
+                    reason.map_or("".to_string(), |reason| format!(", {}", reason)),
+                    c_expected = style::fg(style::Color::Color256(35))
+                        + &style::bg(style::Color::Gray { shade: 3 })
+                        + style::bold(),
+                    c_none = style::reset()
+                )),
+                c_heading = style::fg(style::Color::Color256(172)) + style::bold(),
+                c_none = style::reset()
+            );
+
+            for PackageOutcome { pkg: Package { parsed: pkg, .. }, value } in outliers {
+                report.push_str(&format!(
+                    "\n \u{2022} {c_name}{}{c_none} v{} {c_path}({}){c_none}{}",
+                    pkg.name,
+                    pkg.version,
+                    pkg.manifest_path.strip_prefix(&workspace.root).unwrap(),
+                    value.map_or("".to_string(), |v| format!(
+                        " [found: {c_found}{}{c_none}]",
+                        v,
+                        c_found = style::fg(style::Color::White)
+                            + &style::bg(style::Color::Red)
+                            + style::bold(),
+                        c_none = style::reset()
+                    )),
+                    c_name = style::fg(style::Color::Color256(39)) + style::bold(),
+                    c_path = style::fg(style::Color::Gray { shade: 12 }),
+                    c_none = style::reset(),
+                ));
+            }
+            eprintln!("{}", report);
+            Ok(true)
+        }
+        _ => Ok(false),
+    }
+}
+
+pub fn is_publishable(pkg: &Package) -> bool {
+    !matches!(pkg.raw["package"].get("publish"), Some(toml::Value::Boolean(false)))
+}
+
+/// Returns true if a crate has either of the provided files.
+///
+/// ## Example
+///
+/// ```rust
+/// exists!(pkg, "a" || "b" || ( "c" && "d" ) || "e")
+/// ```
+///
+/// If the crate has either `a`, `b`, `e` or both `c` and `d`, this returns true.
+/// Only returns false if all the combinations don't exist.
+macro_rules! exists {
```

Comment:
I would recommend using a regular `Rust` function. This would be so much easier to review.

Reply:
meh..

```rust
exists!(pkg, "file1" || "file2" || ("file3" && "file4") || "file5")
```

at best, would end up as:

```rust
exists(pkg, "file1") || exists(pkg, "file2") || (exists(pkg, "file3") && exists(pkg, "file4")) || exists(pkg, "file5")
```

Reply:
+1 to what @pmnoxx says. 

Looking at the second example, I can see what does it do. Looking at the first example, I have no idea what's going on, because the stuff after `!` is just custom made up syntax. So I have to either

a) guess intention of the macro writer judging by just how the surface syntax looks. 
b) carefully read the macro definition
c) ask the IDE to expand the macro and read the expansion.